### DISCUSSION TITLE
tutorials/scripting/resources

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#132] [User-contributed notes]
# tutorials/scripting/resources

Nodes and resources: Up to this tutorial, we focused on the Node class in Godot as that's the one you use to code behavior and most of the engine's features rely on it. There is another datatype th...

https://docs.godotengine.org/en/stable/tutorials/scripting/resources.html

**Comment by corjohnson** [comment]
Another word of warning related to how `Resource` are serialized, storing the path of the script used. If you have persisted a resource to the file system (such as a `"user://"` path), the existing persisted data will be corrupt. So if you intend to refactor your custom resource into a new folder, such as moving a script from `res://scripts/user_profile.gd` to `res://data/user_profile.gd`; any instances of the `UserProfile` that have been saved to disk will no longer be able to load.

If you're in development and have not shipped a version of your project yet, you can delete the User Data with the corrupt persisted resource by navigating to `Project > Open User Data Folder`, and then deleting the impacted files.

**Comment by tcm151** [comment]
When using C#, then the `Resource` needs to be in it's own file with the same name as the class, otherwise it will not show up in the create resource GUI.

**Comment by optimized-riyan** [comment]
In C#, if you will be using a custom resource in a node that will be running in the editor (through `[Tool]`), make sure to add `[Tool]` attribute to the custom resource class too, otherwise you will get an invalid cast exception that says `Unable to cast object of type 'Godot.Resource' to type '<Namespace>.<CustomResourceClass>'`.
For example:
```
[GlobalClass][Tool]
public partial class TextureData : Resource {...}
```

**Reply by Leroymilo** [reply]
This can also fix issues in GDScript: I had a custom Resource I used in a tool and couldn't figure out why the getter and setter methods on its properties weren't triggering. Simply adding `@tool` to the custom Resource script fixed the errors.

**Reply by kateonbxsh** [reply]
YOU'RE A LIFE SAVER

**Reply by paulovbettio** [reply]
You can also use this nice syntax sugar for attributes:

```csharp
[GlobalClass, Tool]
public partial class TextureData : Resource {...}
```

**Comment by EdGeber** [comment]
Resources need not be used only as data containers. They can also provide nodes with on-demand algorithm implementations. That is, Resources can be used to implement the [strategy pattern](https://refactoring.guru/design-patterns/strategy).

**Comment by MichaelJBerk** [comment]
When exporting a game, resource files are often renamed to `.tres.remap` (see [66041](https://github.com/godotengine/godot/issues/66014)), which can cause issues when loading resource files from disk. Should this behavior be documented?

**Comment by kora-bert** [comment]
I exit the scene and the resources that were loaded are not freed. I even exit the game (on Android) and reopen it and I still haven't entered the scene that has the resources, but the memory usage is still high due to the resources of that scene. And using CACHE_MODE_IGNORE and CACHE_MODE_IGNORE_DEEP during load doesn't help.

**Comment by Lambach18** [comment]
As i strugled multiple Times with this, in order to use saved resources you need to have a simple constructor, else the scripts don't get atached to the resource files, so make sure .new() is a valid constructor to solve this problem

**Comment by RevolNoom** [comment]
Also note that, if you want a member variable of the Resource to be saved on disk, you need to annotate it with @export. I forgot to add @export, so when I tried to ResourceSaver.save() my resource with plugin, the member variables become empty.

**Reply by Fireye04** [reply]
handy little addendum: if you wish to save/serialize an object without exposing it to the editor, the `@export_storage` flag does so :)

**Comment by torakoneko** [comment]
when i try to use @export in a script that extends Resource, i get this error
```
Error at (3, 1): Node export is only supported in Node-derived classes, but the current class inherits "Resource".
```
which appears to be a fix for a bug that did allow that, but this tutorial suggests doing that so i'm kind of confused.

**Reply by christopher-r-brech** [reply]
I'm having this problem with Godot 4.7. Everything I find seems to indicate using @export in a Resource is perfectly normal, but I get this same error when I try to do it, and it's not mentioned anywhere else that I can find that it's a problem.

**Reply by christopher-r-brech** [reply]
Can't seem to edit my message above, but the error message wasn't clear what the actual problem was. It's not that you can't use @export in anything but classes that inherit from Node, it's that a Resource can't use @export with Node-derived classes. So `@export var a_vector : Vector2i` is fine, but `@export var a_tileMapLayer : TileMapLayer` is not.

I was trying to use a Resource because I was led there after trying to export an array of an inner class and discovered that's a no-go.

**Comment by tgbrow** [comment]
In the `Bot` code snippet in the "Creating your own resources" section, I'm curious if there's a compelling reason why `stats` is labelled with type `Resource` rather than `BotStats`.

In other words, why this:
```
@export var stats: Resource
```
instead of this?
```
@export var stats: BotStats
```

With the latter option, it would be guaranteed that `stats.health` exists.

**Reply by KANNAPE** [reply]
It's because GDScript allows duck-typing. In other words, this code will work whether you use a BotStats type resource, or another one, as long as both have a health property. In the long term, if you want to add, let's say, a new resource called HumanStats, or EnemyStats, or whatever you want, you won't need to specify for each case what to do, the script will automatically know what variable to look for.
Hope it'll help!

**Comment by Peppe1223** [comment]
Sorry, but I'm a little confused about one thing, namely why I should insert a method into a resource if, in the end, I can insert that method into a scene and then retrieve the resource data from the scene, i.e., from the method I defined within it?

**Reply by Bkjvk** [reply]
Well you dont have to insert any methods into a resource, from the code perspective it's just another object, so if you can put some logic there that would simplify it's usage that's your go to. Let's say we have some: 
```C#
public partial class Stats : Resource
{
    [Export] public int Health;
    [Export] public int Shield;
}
```
You would take some damage, then instead of inserting the code for taking in onto your player code you insert it inside of resource aka:
```C#
public partial class Stats : Resource
{
    ...exports...

    public bool TakeDamage(int damage)
    {
        if (Shield >= damage)
        {
             Shield -= damage;
             return false;
        }
        damage -= Shield;
        Shield = 0;
        Health -= damage;
        return Health <= 0 ? false : true;
}
```
There is additional benefit to it, your oponent will have different Script than your player and they dont need to implement `TakeDamage` method as long as they also use `Stat` resource.
Does it answer your question?

**Reply by Peppe1223** [reply]
Okay... what you mean is sharing a resource, more precisely that I can call that method you defined in the resource from another scene? The point I want you to understand is that defining a method in a resource when you know in advance that that method will only be used for one scene and one scene only (the situation would change if that method were to be used in more than one scene, so I would share the resource to use the method in other scenes) is pointless... or I can understand if you put this method in the same resource for organizational purposes and therefore simply to contain the method... But I could be wrong... Tell me if I'm wrong

**Reply by Bkjvk** [reply]
Its not stupid, If I used the `Stat` object just in `Player` scene its still good because player object usually have a lot of code so moving that code to different object can increase readability. If the scene that's using Resource doesnt contain that much code you can keep the code within the scene because it doesnt matters. Its up to you as developer to decide how you wanna organize your code.

**Comment by UrbanNuke** [comment]
It there any way to fill description for custom resources using C#?

**Comment by gnat** [comment]
For people wondering about Resources vs Scenes.

* Resource = No Scene Tree stuff.
* Scene = Resource with Scene Tree stuff.

In practice, real workflow differences comparing the two:

Inspector Window (Example: `@export var cosmetic:Cosmetics`):

* Resource: I can choose Cosmetics Resources on the File System.
* Scene: I can choose Cosmetics Scenes from the Scene Tree.
  * With `PackedScene` I can choose from File System but I CANNOT select Cosmetics Scenes specifically.

Instances:

* Resource: Instantiated once, instance is referenced everywhere.
  * Can be instantiated with `new()`
* Scene: Instantiated many times, instance is used directly.

**Reply by gnat** [reply]
Also `preload()` returns a `PackedScene` for `.tscn`, so in practice you basically need to `.instantiate()`

Resources, you can just use the `preload()` instance as-is.
